pkg/controllers/mpi_job_controller.go

```diff
@@ -480,13 +475,36 @@ func (c *MPIJobController) getLauncherJob(mpiJob *kubeflow.MPIJob) (*batchv1.Job
 	return launcher, nil
 }
 
-// getTotalGPUs gets the total number of desired GPUs. Defaults to 1 if not specified.
-func getTotalGPUs(mpiJob *kubeflow.MPIJob) int {
-	totalGPUs := 1
+// allocateGPUs allocates the worker replicas and GPUs per worker.
+func allocateGPUs(mpiJob *kubeflow.MPIJob, gpusPerNode int, done bool) (workerReplicas int, gpusPerWorker int, err error) {
+	workerReplicas = 0
+	gpusPerWorker = 0
+	err = nil
 	if mpiJob.Spec.GPUs != nil {
-		totalGPUs = int(*mpiJob.Spec.GPUs)
+		totalGPUs := int(*mpiJob.Spec.GPUs)
+		if totalGPUs < gpusPerNode {
+			workerReplicas = 1
+			gpusPerWorker = totalGPUs
+		} else if totalGPUs % gpusPerNode == 0 {
+			workerReplicas = totalGPUs / gpusPerNode
+			gpusPerWorker = gpusPerNode
+		} else {
+			err = fmt.Errorf("specified #GPUs is not a multiple of GPUs per node (%d)", gpusPerNode)
+		}
+	} else if mpiJob.Spec.Replicas != nil {
+		workerReplicas = int(*mpiJob.Spec.Replicas)
+		container := mpiJob.Spec.Template.Spec.Containers[0]
+		if container.Resources.Limits != nil {
```

Comment:
If the limit is not set, `gpusPerWorker` would take the default value of 0.

Reply:
Ok, will you continue to execute if number of GPU specified not equal to number of GPUs per node?

Reply:
Right now the answer is yes. When the user specifies `replicas` and the GPU resource limit explicitly, they are basically on their own. I can imagine in large clusters it might be useful to have this flexibility. For example, if your job's communication overhead is not too big, and half nodes are a lot easier to schedule than full nodes, maybe you'd want to ask for 4 GPUs instead of 8?

Reply:
I guess the more obvious use case is a mixed cluster where you have nodes with different number of GPUs. We can come up with some elaborate scheme to allocate GPUs optimally, but as a first step should probably just give users the flexibility.

Reply:
Ok, if you are working with MPI using TCP than it is ok. But for MPI with infiniband would need IB device association with POD, probably thats not the focus here. Looks good to me